### PR TITLE
WIP: mask geopandas

### LINKED ISSUE #102: mask geopandas dataframe directly?
Instead of converting a GeoDataFrame into a structure that works for regionmask it should be possible to just build a thin wrapper around `_mask` which takes a GeoDataFrame as input.

### PR DESCRIPTION
**Pull request by mathause**
- [x] Closes #102
 - [ ] Tests added
 - [ ] Passes `isort -rc . && black . && flake8`
 - [ ] Fully documented, including `whats-new.rst`

This would be an easy way to mask a GeoDataFrame directly without the detour to the `regionmask.Regions`.

@aaronspring

### CODE REVIEW

**Comment by mathause** [comment]
@aaronspring

The code in `regionmask/core/mask_geopandas.py` should be copy-pastable, then this example should work:

```python
from cartopy.io import shapereader
import geopandas
import numpy as np

resolution="110m"
category="cultural"
name="admin_0_countries"

shpfilename = shapereader.natural_earth(resolution, category, name)
df = geopandas.read_file(shpfilename, encoding="utf8")

lon = np.arange(0.5, 359.5)
lat = np.arange(89.5, -89.5, -1)

mask = mask_geopandas(df, lon, lat)

mask.plot()
```

**Comment by aaronspring** [comment]
The PR works on my machine.

**Comment by aaronspring** [comment]
This provides a xr object which has increasing integer values for the different geodf polygons. Works also in my machine. And on my mpiom curvilinear grid

File: regionmask/core/mask.py [hunk 1]
```diff
@@ -99,18 +52,15 @@ def _mask(
...
-        data = self.coords
     elif method == "rasterize":
         func = _mask_rasterize
-        data = self.polygons
     elif method == "shapely":
         func = _mask_shapely
-        data = self.polygons
     else:
         msg = "Only methods 'rasterize', 'shapely', and 'legacy' are implemented"
         raise NotImplementedError(msg)
```
Comment (mathause) [inline regionmask/core/mask.py:61]:
```suggestion
        msg = "Method must be one of 'rasterize', 'shapely', or 'legacy'."
        raise ValueError(msg)
```

File: regionmask/core/_geopandas.py [hunk 2]
```diff
@@ -52,7 +52,7 @@ def _construct_abbrevs(geodataframe, names):
             "geodataframe, choose from {}".format(geodataframe.columns)
         )
     abbrevs = []
-    names = _maybe_get_column(geodataframe, names)
+    names = geodataframe[names]
```
Comment (mathause) [inline regionmask/core/_geopandas.py:55]:
```suggestion
    _maybe_get_column(geodataframe, names)
```

File: regionmask/core/_geopandas.py [hunk 3]
```diff
@@ -91,13 +91,15 @@ def from_geopandas(
...
+    from geopandas import GeoDataFrame
+
+    if not isinstance(geodataframe, (GeoDataFrame)):
+        raise TypeError("`geodataframe` must be a geopandas 'GeoDataFrame'")
 
     if numbers is not None:
         # sort, otherwise breaks
         geodataframe = geodataframe.sort_values(numbers)
-        numbers = _maybe_get_column(geodataframe, numbers)
+        numbers = geodataframe[numbers]
```
Comment (mathause) [inline regionmask/core/_geopandas.py:102]:
```suggestion
    _maybe_get_column(geodataframe, names)
```

File: regionmask/core/_geopandas.py [hunk 4]
```diff
@@ -106,19 +108,19 @@ def from_geopandas(
     numbers = np.array(numbers)
 
     if names is not None:
-        names = _maybe_get_column(geodataframe, names)
+        names = geodataframe[names]
```
Comment (mathause) [inline regionmask/core/_geopandas.py:111]:
```suggestion
    _maybe_get_column(geodataframe, names)
```

File: regionmask/core/_geopandas.py [hunk 5]
```diff
@@ -106,19 +108,19 @@ def from_geopandas(
...
         _check_missing(names, "names")
         _check_duplicates(names, "names")
 
     if abbrevs is not None:
         if abbrevs == "_from_name":
             abbrevs = _construct_abbrevs(geodataframe, names)
         else:
-            abbrevs = _maybe_get_column(geodataframe, abbrevs)
+            abbrevs = geodataframe[abbrevs]
```
Comment (mathause) [inline regionmask/core/_geopandas.py:119]:
```suggestion
    _maybe_get_column(geodataframe, names)
```

File: regionmask/core/regions.py [hunk 6]
```diff
@@ -256,12 +256,81 @@ def lon_360(self):
...
+        method=None,
+        xarray=None,
+        wrap_lon=None,
+    ):
+        """
+        create a grid as mask of a set of regions for given lat/ lon grid
+
+        Parameters
+        ----------
+        lon_or_obj : array_like or object
```
Comment (mathause) [inline regionmask/core/regions.py:274]:
```suggestion
        lon_or_obj : object or array_like
```

File: regionmask/core/_geopandas.py [hunk 7]
```diff
@@ -128,3 +130,95 @@ def from_geopandas(
...
+    wrap_lon=None,
+):
+    """
+    create a grid as mask of a set of regions for given lat/ lon grid
+
+    Parameters
+    ----------
+    geodataframe : GeoDataFrame or GeoSeries
+        Object providing the region definitions (outlines).
+    lon_or_obj : array_like or object
```
Comment (mathause) [inline regionmask/core/_geopandas.py:152]:
```suggestion
    lon_or_obj : object or array_like
```

File: regionmask/test/test_geopandas.py [hunk 8]
```diff
@@ -165,3 +172,70 @@ def test_construct_abbrevs():
     expected = ["A", "Bce", "Ghi", "Jk", "Lmn", "OpQr", "StuXyz"]
     for i in range(len(result)):
         assert result[i] == expected[i]
+
+
+# ==============================================================================
+
```
Comment (mathause) [inline regionmask/test/test_geopandas.py:178]:
```suggestion
# uses the same function as `Regions.mask` - only do minimal tests here
```

**Pull request merged**